numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,745 @@
+            }
+        }
+        unpacked_string->size = size;
+        unpacked_string->buf = buf;
+    }
+
+    return 0;
+}
+
+static char *
+heap_or_arena_allocate(npy_string_allocator *allocator,
+                       _npy_static_string_u *to_init_u, size_t size,
+                       int *on_heap)
+{
+    unsigned char *flags = &to_init_u->direct_buffer.size_and_flags;
+    if (*flags & NPY_STRING_SHORT) {
+        // Have to heap allocate since there isn't a preexisting
+        // allocation. This leaves the NPY_STRING_SHORT flag set to indicate
+        // that there is no room in the arena buffer for strings in this
+        // entry
+        *flags |= NPY_STRING_ON_HEAP;
+        *on_heap = 1;
+        return allocator->malloc(sizeof(char) * size);
+    }
+    npy_string_arena *arena = &allocator->arena;
+    if (arena == NULL) {
+        return NULL;
+    }
+    if (*flags & NPY_STRING_ARENA_FREED) {
+        // Check if there's room for the new string in the existing
+        // allocation. The size is stored one size_t "behind" the beginning of
+        // the allocation.
+        char *buf = vstring_buffer(arena, to_init_u);
+        if (buf == NULL) {
+            return NULL;
+        }
+        size_t alloc_size;
+        if (is_medium_string(to_init_u)) {
+            // stored in a char so direct access is OK
+            alloc_size = (size_t) * (buf - 1);
+        }
+        else {
+            // not necessarily memory-aligned, so need to use memcpy
+            size_t *size_loc = (size_t *)((uintptr_t)buf - sizeof(size_t));
+            memcpy(&alloc_size, size_loc, sizeof(size_t));
+        }
+        if (size <= alloc_size) {
+            // we have room!
+            *flags &= ~NPY_STRING_ARENA_FREED;
+            return buf;
+        }
+        else {
+            // No room, resort to a heap allocation.
+            *flags = NPY_STRING_ON_HEAP;
+            *on_heap = 1;
+            return allocator->malloc(sizeof(char) * size);
+        }
+    }
+    // string isn't previously allocated, so add to existing arena allocation
+    char *ret = arena_malloc(arena, allocator->realloc, sizeof(char) * size);
+    if (size < NPY_MEDIUM_STRING_MAX_SIZE) {
+        *flags |= NPY_STRING_MEDIUM;
+    }
+    return ret;
+}
+
+static int
+heap_or_arena_deallocate(npy_string_allocator *allocator,
+                         _npy_static_string_u *str_u)
+{
+    unsigned char *flags = &str_u->direct_buffer.size_and_flags;
+    if (*flags & NPY_STRING_ON_HEAP) {
+        // It's a heap string (not in the arena buffer) so it needs to be
+        // deallocated with free(). For heap strings the offset is a raw
+        // address so this cast is safe.
+        allocator->free((char *)str_u->vstring.offset);
+        if (*flags & NPY_STRING_SHORT) {
+            *flags = 0 | NPY_STRING_SHORT;
+        }
+        else {
+            *flags &= ~NPY_STRING_ON_HEAP;
+        }
+    }
+    else if (VSTRING_SIZE(str_u) != 0) {
```

Comment:
The empty string is special because we want it to have a packed string representation as a zero-filled 16 byte buffer (on 64 bit systems). That means the flags are unset, so it's not a short string, which means calling `NpyString_free(empty_string)` gets to this if statement. Added a comment explaining.

Reply:
Part of the reason I later wondered about `LONG_STRING` as an indicator instead of `SHORT_STRING`: good to have all-zero be valid on its own accord.